### PR TITLE
Fix problem that may have made 'Hello World' print more than once

### PR DESCRIPTION
**Pull request by mpetch**
Fix problem that may have made 'Hello World' print more than once or ended in unusual behavior. Ensure CLI is used before calling HLT so that interrupts don't restart the code after the HLT. Also ensure that when updating SS and SP that interrupts are off during the operation. It is possible for an interrupt to occur between the instructions causing the stack state to be invalid during interrupt execution,

### CODE REVIEW

File: hajji/hello.S [hunk 1]
```diff
@@ -49,8 +48,15 @@ start:
          * we anchor it at $LOAD.  Note that we don't collide with
          * the code because the stack will always remain below
          * (i.e. less than) $LOAD and grows downwards from there.
+         * disable intterupts when setting up the stack. If an
+         * interrupt occurs between the two MOVs then the stack
+         * may point at the wrong memory location and the interrupt
+         * may crash the system
          */
+        cli
```
Comment (cirosantilli) [inline hajji/hello.S]: Why not do this as the very first instruction of the program?

File: hajji/hello.S [hunk 2]
```diff
@@ -49,8 +48,15 @@ start:
...
+        movw %ax, %ss
         movw $LOAD, %sp
+        sti
```
Comment (cirosantilli) [inline hajji/hello.S]: Why is this needed?

File: hajji/hello.S [hunk 3]
```diff
@@ -61,6 +67,7 @@ start:
         callw clrscr                  # clear screen
         callw curshome                # move cursor home - top:left
         callw greeting                # display a greeting string
+		cli
```
Comment (cirosantilli) [inline hajji/hello.S]: This is likely a typo since the next instruction is also a cli.

**Comment by cirosantilli** [comment]
Thanks for the commit, it did work now.

Please break the commit message into a short title line and a body: http://chris.beams.io/posts/git-commit/ ;-)

File: hajji/hello.S (hunk 1 above)
Comment (mpetch) [inline hajji/hello.S]: You can do it as the first line of the program however depending on what code someone is going to write they may want want to run with interrupts on for performance reasons. For small programs that don't use interrupts then turning them off at the start is fine. If you use interrupts it is good practice to keep them off for the minimum time necessary and re-enable them as quickly as possible.

Something else you should do is specifically when your program starts is call "CLD" if you intend to use any of the looping constructs. You can't guarantee the direction flag will be set properly.

File: hajji/hello.S (hunk 2 above)
Comment (mpetch) [inline hajji/hello.S]: See my other comment about performance and running with interrupts on. If you aren't using interrupts in an application then turn them off at the start. Be aware that putting CLI and STI around the stack loading code generally wouldn't be an issue because the processor is suppose to keep interrupts off while changing a segment register and re-enable them automatically after the following instruction. If running on an old bare metal system with an ancient 8088 processor it is possible you have a buggy chip that doesn't honor this.

File: hajji/hello.S (hunk 3 above)
Comment (mpetch) [inline hajji/hello.S]: It wasn't a typo since I wrote the code to turn interrupts off only when necessary and enable them when appropriate to do so.

Comment (mpetch) [inline hajji/hello.S]: Copy and paste without removing the original. I originally had it outside the comment and intended to paste it below the comment after but didn't remove the original.

**Comment by cirosantilli** [comment]
Another tip: use feature branches. Note how this PR got cluttered with the commit on the other PR because both are from branch `master` :-)

**Comment by cirosantilli** [comment]
Merged at: https://github.com/cirosantilli/x86-bare-metal-examples/commit/617af66caef16a018e07ac85ed83208b96675958
